### ISSUE TITLE
Performance problem when starting first time

### ISSUE THREAD

**Issue by klartext** [newsboat/newsboat#766]
Version-Info
```
newsboat 2.18.0 - https://newsboat.org/
Copyright (C) 2006-2015 Andreas Krennmair
Copyright (C) 2015-2019 Alexander Batischev
Copyright (C) 2006-2017 Newsbeuter contributors
Copyright (C) 2017-2019 Newsboat contributors

Newsboat is free software licensed under the MIT License. (Type `newsboat -vv' to see the full text.)
It bundles JSON for Modern C++ library, licensed under the MIT License: https://github.com/nlohmann/json

newsboat 2.18.0
System: Linux 5.5.3-arch1-1 (x86_64)
Compiler: g++ 9.2.0
ncurses: ncurses 6.1.20180127 (compiled with 6.1)
libcurl: libcurl/7.68.0 OpenSSL/1.1.1d zlib/1.2.11 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh2/1.9.0 nghttp2/1.39.2 (compiled with 7.67.0)
SQLite: 3.31.1 (compiled with 3.30.1)
libxml2: compiled with 2.9.10
```

Config:
```
browser firefox -private
show-read-feeds no
save-path  ~/newsbeuter-extracted
cleanup-on-quit yes
```
I have roughly about 500 feed-urls in my urls-file.
The site.db is now round about 7 GB large.
First start takes between 5 and 6 minutes.
quitting and starting again and newsboat available in less than about 15 seconds.

I would guess that newsboat's allocation of memory for feeds is done without such a huge urls and site.db file in mind. (For second start, linux kernel might do caching, so it comes up faster.)
Would it be possible to enhance the performance here, by possibly changing the memory allocation, to gain a speedup here?

**Comment by Minoru** [comment]
Yes, I've been thinking about this problem for a while now, but haven't found a solution yet. One part of the problem is described in #587. Another is that Newsboat *has* to read the whole DB on startup in order to apply killfiles. It's easy to fix in a backward-compatible manner (just add a new column, a boolean flag indicating if the article is ignored), but I want forward-compatibility too, and don't see how to achieve it.

**Comment by klartext** [comment]
if db_version < x then use old_bahviour else use_new_behaviour.

But, oh, looks like the db itself has no version-information?!

Then analyzing the schema: if column exists, use new behaviour, else use old behaviour?
Old code just uses old behaviour, because that column is not known. (Should be that way, at least - depends on the queries I guess).

btw.: it's only 2 GB, which takes 5...6 minutes; the 7GB was from one of my own tools, which I confused with the newsboat cache.db

**Comment by klartext** [comment]
There is a db_schema_version_major (and minor) in metadata-table. Couldn't that be used for decisions on startup?

**Comment by Minoru** [comment]
> if db_version < x then use old_bahviour else use_new_behaviour

That only provides backwards compatibility, but I want forward one as well.

Some users sync their cache.db between computers, and there is no guarantee that all of those computers will switch to a newer version simultaneously. It's entirely possible to end up in a situation where:
1. newer version starts, and upgrades the DB. It adds a new column, and updates it accordingly
2. older version opens an already-updated DB. It doesn't know anything about the new column, so it never updates it
3. newer version starts again. It sees there is a new column, and trusts it blindly—even though the column is now out of sync

I came up with a few ideas, but none of them would work:
1. **make newer version re-scan the DB to update the column**. This gets us back to square 1—Newsboat will re-read the whole DB on each start
2. **use an SQL trigger to note when the DB is opened**. But that's just not possible in SQLite: all triggers should attached to some query.

That's about as far as my thinking got so far. If you have other ideas, let's hear them out!

**Comment by klartext** [comment]
If there would be no killfiles, what then?
Would read newsboat the feeds just when entering a feed?
Is only the killfile-issue the problem?
If so, there could be a check, if killfiles are used at all. (I don't use them, so a check would be of no help at all.)
So, if killfiles are not used, the check would not be necessary.

If you use the killfile thing via database, why using a boolean for checked?
Is the killfile-decision done always when starting, during runtime?
Is there no entry for this in the database at all?

There could also be a seperate table for this, instead of a new column.
The killfile-check could then be done once per rss-entry.
The IDs of the entries could be used to identify the checked items.
And a boolean says, if that entry is in killfile or not.
The maximum of the IDs in the killfile-checker-table would indicate, if a message must be read and checked or not. The old newsboat would just ignore that (and read all the data). But the new newsboat would look up max ID in that table to see, what has been checked and what not.
If there are messages/entries with ID > max_ID_of_checker_Table, then the data must be read, and the db could be updated then.


I think that, but I'm not sure if  that works; just some thoughts about that. I have not looked into newaboat's code and don't know the logic of how the data is handled. So, maybe I have overlooked something. But maybe it makes sense and can be applied?

**Comment by Minoru** [comment]
Hmm, your questions made me re-read [`Cache::internalize_rssfeed()`](https://github.com/newsboat/newsboat/blob/b8c7a2ee5ab73474a0f09b0ef239205b8785dcad/src/cache.cpp#L562), and it doesn't actually read the contents of the articles. I might be misunderstanding what's causing the slowdown.

> If there are messages/entries with ID > max_ID_of_checker_Table, then the data must be read, and the db could be updated then.

That's a clever idea, but it breaks when an item is updated. An update could make an item match some killfile. The old version won't update the table, and the new version won't know it has to re-scan the DB.

Now, `UPDATE` is something we could attach a trigger to, but that'd amplify I/O on each update. Meh.

**Comment by klartext** [comment]
If an item is updated (I think, you mean the feed itself has changed), then the "checked-flag" must be reset for that item, or the killfile-check must be done imediately, when updating.
Don't see, why there must be a trigger.

**Comment by Minoru** [comment]
> If an item is updated (I think, you mean the feed itself has changed)

Yeah, the item in the feed changed, e.g. an article was published and later it was corrected to fix typos, or there was some addendum.

> then the "checked-flag" must be reset for that item, or the killfile-check must be done imediately, when updating

Older version don't do this. Sure, we can teach new versions to do that, but that only gives us backward compatibility, not forward one.

A trigger would let us circumvent the problem, because the logic will be stored in the DB itself, and would be applied regardless of Newsboat version that uses the DB.

**Comment by klartext** [comment]
OK, then there must be the date of the last check.
```entry_id, datetime_check, status```

Then lookup the datetime of the entry and compare it with the datetime of the last check. if the entry has changed, it's datetime is newer than the last-check-datetime. Then read the whole entry and check if it must be killed.
Of course, if there is no killfile in use, the whole thing does not need to be started.
And if the old newsboat behaves the old way, it's not worse. The new newsboat could save time then.

**Comment by Minoru** [comment]
> Then lookup the datetime of the entry

That date is taken from the feed, and the feed is free to change it any way it likes—e.g. it can move it into the past, which will get around the check you suggest. Also, there are no guarantees that one of user's computers don't have the wrong date and time.

**Comment by Oppen** [comment]
@klartext do you have plenty of RAM? Because that would explain why you only have that use on the first start: Linux has a page cache where it stores recently accessed and speculated to be soon accessed file contents. Once it was loaded the first time, if you have enough RAM it will stay in memory until there's enough memory pressure for it to become a problem and be discarded.
I know this doesn't fix or change the issue, but it's just a hint to understand it a bit better.

**Comment by klartext** [comment]
> @klartext do you have plenty of RAM? Because that would explain why you only have that use on the first start: Linux has a page cache where it stores recently accessed and speculated to be soon accessed file contents. Once it was loaded the first time, if you have enough RAM it will stay in memory until there's enough memory pressure for it to become a problem and be discarded.
> I know this doesn't fix or change the issue, but it's just a hint to understand it a bit better.

I know this cache-thing.
And "plenty of RAM" ... plenty compared to what? Maybe enough to read the index data from sqlite-db, but the used file was so huge (7 GB), it would eat 7/8  of my 8GB RAM, if it would be read completely.
I think only some metadata to the db is read. And even that just takes too long.

I found a workaround: splitting the database and the url files into seperate chunks: each one for a certain topic. It's easier to find what I'm looking for (alias to the newsboat-calls with it's seperated db- and url-file) , and it speeds up things (some db-files are some MB, one is about 250 MB). Nevertheless for remaininf old the 2 GB cache.db (formerly 7 GB) it's slow.

The problem with that workaround-approach: the old data is in the old db-file, and if there is no way to migrate it to another newsboat database-instance, I can't remove it from the old database. Data would be lost. I might import that data into my own database - a while ago I have written my own tool that reads out newsbeuter/newsboat-data and inserts it into my own link-database. So I may use that way. But my link-db does contain less fields, so some data might vanish, so only title, date and URL would be left over. That's ok for some of the feeds. But for some podcast feeds I would like to have them complete in newsboat, to look up old items, and have them all in one place.

Hmhhh, maybe I should write my own curses-based rss-collector tool? ;-)

**Comment by Oppen** [comment]
"Plenty of RAM" means "most of the file fits your free RAM". Note I'm 100% against the "but RAM is cheap" argument, it kinda assumes you live in a relatively high-income household in the first world. I was just guessing why the issue seems to appear on first start and doesn't repeat itself afterwards. That obviously doesn't change the fact that there's an issue to fix.

That workaround is pretty smart, BTW.

Regarding migration, there's probably a not-too-hard way to `SELECT` and insert to a different DB.

**Comment by Minoru** [comment]
If you're going to discuss migration, please either open a new issue, or move to the mailing list.

**Comment by klartext** [comment]
> If you're going to discuss migration, please either open a new issue, or move to the mailing list.

Is that already planned (inofficially, without issue), to support that?

**Comment by Minoru** [comment]
@klartext No, and it seems like a one-off thing, so I don't plan to support that.

**Comment by allanwind** [comment]
`Loading articles from cache...` takes about ~5 sec for me with 1227 feeds and `keep-articles-days 365` on a fast nvme drive.  My cache.db is 49MB.  I `keep-articles-days 365` so I can identify zombie feeds (feeds that did not publish any articles for a year) so they be pruned from .newsboat/urls.  This means I cannot use `delete-read-articles-on-quit yes` even though I don't really care about read articles older than a day or two.  For me `keep-articles-days 365` is a work-around for two minor missing features a) keep stats around to delete zombie feeds (this could be as simple as suitable log), and b) delete-read-article-days.  It may be entirely different for @klartext of course.

What about lazy loading the cache per feed?  For me maybe 2% of the feeds have new articles on a daily run and that usually the only feeds that I care about (counter example search?).  And even for the feeds with new articles, if `show-read-articles no` the unread articles would not be needed other than to merge the (remote) feed.  You could initiate an eager load in a separate thread if a the latency to open a feed is bad.  If engage in reading a single article the entire eager load time would be entirely hidden.

The `ignore-mode display` is the only safe mode, i.e. not resulting in data loss if you mess up a regex in `ignore-article "*"`.  Iff `ignore-mode download` is blocking progress maybe we should removing that feature?

**Comment by Minoru** [comment]
> What about lazy loading the cache per feed?

Feedlist needs to know how many unread and total articles there are in feed. To get those counts, we have to apply killfiles first. If `ignore-mode` is `display`, it means loading at least a part of the article. So I don't think we can lazy-load.

>  Iff ignore-mode download is blocking progress maybe remove that feature?

`ignore-mode download` is not blocking this; in that mode, ignored articles are never even stored in the cache (not even their GUIDs). It's the display mode that makes this complicated (and also my insistence on forward compatibility).

It might well be that the easiest way forward here is to release Newsboat 3 :) Then we don't have to worry about forward compatibility from old 2.x versions anymore.

**Comment by allanwind** [comment]
> Feedlist needs to know how many unread and total articles there 
> are in feed. To get those counts, we have to apply killfiles 
> first. If `ignore-mode` is `display`, it means loading at least a 
> part of the article. So I don't think we can lazy-load.

The normal case is that ignore-articles filters haven't changed
which would allow you to use cache statistics for number of read 
and unread articles.  This means that you only have to update
the counters for "new" articles.  "update" of cached values, btw,
scares me a little bit in terms of consistency between cache
and baseline data.

The unread count is important but I only care about the total
counts when identifying zombie feeds once in a while.  If you only 
show the unread counts in the feedlist then you only need to
calculate the statistics for the "new" articles if 
`show-read-articles no`.

I have a more radical idea... we could move to a "flat" view that 
contains all articles across feeds.  This would eliminate the 
whole navigation between feeds, and `mark-feed-read` would just 
apply to feed for the current article.  Probably need to support a 
two-level sort, so you can do (feed asc, date desc).  Maybe a 
header/footer thing too.

> It might well be that the easiest way forward here is to release 
> Newsboat 3 :) Then we don't have to worry about forward 
> compatibility from old 2.x versions anymore.

+1

**Comment by Minoru** [comment]
> I have a more radical idea... we could move to a "flat" view that contains all articles across feeds

How doe this fix the performance issue?

I also don't see a reason to change the current, time-tested UI. We already support "river of news" approach through query feeds, and the people who wanted three-pane layout never returned with patches. Other than that, I don't remember any requests to change UI, so I think the current one is good enough. (I'm always open to discussion, but if it's not related to this issue, please open a new one).

**Comment by allanwind** [comment]
On 2021-07-30T14:24:40, Alexander Batischev wrote:
>> I have a more radical idea... we could move to a "flat" view that contains all articles across feeds
>
> How doe this fix the performance issue?

If the view doesn't have to show stats per feed, then you don't 
need to calculate them.  Or said differently optimize for unread 
articles instead of all articles.

I follow a ridiculous 1228 feeds, and it will take time to hit 
those, but that happens in a different thread pool so you can 
start consuming articles as soon as the first feed with new
content is read.  This works really well.  What is not so great
is that it takes ~5s on the main thread to read the cache which 
you have to wait for to complete.  What can we do to make 
time(show first unread article) = time(download first feed with 
unread article)?  Just eye balling it the latter is ~0.1s on
my system.  The other part of that is newsboat is doing ~5s of 
processing to show me 130 new, non-ignored articles.  How can
we make newsboat do less work?

> I also don't see a reason to change the current, time-tested UI. 
> We already support "river of news" approach through query feeds, 
> and the people who wanted three-pane layout never returned with 
> patches. Other than that, I don't remember any requests to 
> change UI, so I think the current one is good enough. (I'm 
> always open to discussion, but if it's not related to this 
> issue, please open a new one).

The UI is great even without the query feature as `n` allows you 
easily view all unread articles across feeds.  If the UI is 
getting in the way of optimizing for unread articles maybe it
is worth considering?

**Comment by Minoru** [comment]
> If the view doesn't have to show stats per feed, then you don't need to calculate them.

But we still have to load articles to list them, so what's the difference?

Lazy-loading and alternative UIs would be appropriate if we hit some fundamental limit that prevented us from going any faster (such as network delay on a single request that we absolutely must do). But I don't think we're hitting that yet. We just have to store a bit more info in the cache.db, so we don't have to re-read the entire thing on each startup.

**Comment by allanwind** [comment]
On 2021-07-31T07:57:07, Alexander Batischev wrote:
>> If the view doesn't have to show stats per feed, then you don't need to calculate them.
>
>But we still have to load articles to list them, so what's the difference?

If `show-read-articles no`, then you only need to process the unread 
articles.  If `show-articles yes` you page through results 
instead of eagerly load them all (i.e. retrieve sorted results)

> Lazy-loading and alternative UIs would be appropriate if we hit 
> some fundamental limit that prevented us from going any faster 
> (such as network delay on a single request that we absolutely 
> must do). But I don't think we're hitting that yet. We just have 
> to store a bit more info in the cache.db, so we don't have to 
> re-read the entire thing on each startup.

Ok.


/Allan
-- 
Allan Wind
Yaxto - Runs Your Business
<https://yaxto.com>

**Comment by yoshi314** [comment]
> I came up with a few ideas, but none of them would work:

from what i understand this might be done like this (i have seen it in a few solutions, and it works pretty okay-ish).

provide table last_db_access with columns of  :

```
schema_version | timestamp
```

where schema_version will start at 1 for now. each major newsboat release that changes the schema will use increased schema_version number for logging into this table. Older newsboat will obviously not log there at all.


each version of newsboat that is aware of schema_version will log there. Only one row per schema_version. If db was touched, update/insert row with your schema_version and current timestamp


This allows for a few checks:

are we the most recent newsboat accessing this db?

(max(lastmodified) from rss_feed  <= timestamp for our schema_version and there is no higher schema_version in the table than ours) 

was the db modified by older newsboat client?  
( max(lastmodified) from rss_feed  is higher than all entries in last_db_access, meaning db was modified by an old newsboat app that was unaware of this table ) 


if we are certain that newsboat with our schema_version recently used the db - keep running as-is.

if db was touched by older newsboat - do some housekeeping. (that may vary depending on the version of newsboat that last touched the db, more or less work will be required. also, depending on how many schema revisions happened inbetween).

if there was a newer newsboat release accessing the db - debatable. i would assume, run as-is and newer version will further fix the db entries, when it detects the changes. 


Also, the indexes from related bug report help a LOT with startup and general operation. It's worth introducing them even now. (in my case they cut down startup from 8s to under 0.5s)
